solver/src/main/java/org/chocosolver/solver/search/loop/monitors/NogoodFromRestarts.java

```diff
@@ -46,107 +52,110 @@
-                            png.addLearnt(Arrays.copyOf(lits, i + 1));
-                        }
-                    }
-                } else {
-                    throw new UnsupportedOperationException("NogoodStoreFromRestarts cannot deal with such operator: " + ((IntDecision) decision).getDecOp());
-                }
-            } else if (decision instanceof SetDecision) {
-                SetDecision sd = (SetDecision) decision;
-                if (sd.getDecOp() == DecisionOperatorFactory.makeSetForce()) {
-                    if (sd.hasNext() || sd.getArity() == 1) {
-                        lits[i++] = SatSolver.negated(png.Literal(sd.getDecisionVariable(), sd.getDecisionValue(), true));
-                    } else {
-                        if (i == 0) {
-                            // value can be removed permanently from var!
-                            png.addLearnt(SatSolver.negated(png.Literal(sd.getDecisionVariable(), sd.getDecisionValue(), true)));
-                        } else {
-                            lits[i] = SatSolver.negated(png.Literal(sd.getDecisionVariable(), sd.getDecisionValue(), true));
-                            png.addLearnt(Arrays.copyOf(lits, i + 1));
-                        }
-                    }
+            int lit = asLit(decision);
+            if (decision.hasNext() || decision.getArity() == 1) {
+                lits[i++] = lit;
+            } else {
+                if (i == 0) {
+                    // value can be removed permanently from var!
+                    png.addLearnt(lit);
                 } else {
-                    throw new UnsupportedOperationException("NogoodStoreFromRestarts cannot deal with such operator: " + ((SetDecision) decision).getDecOp());
+                    lits[i] = lit;
+                    png.addLearnt(Arrays.copyOf(lits, i + 1));
                 }
-            } else {
-                throw new UnsupportedOperationException("NogoodStoreFromRestarts can only deal with IntDecision and SetDecision.");
             }
         }
     }
+
+    /**
+     * Transform this decision into a literal to be used in {@link PropNogoods}.
+     *
+     * @param decision a decision
+     * @return the literal corresponding to this decision
+     */
+    private <V extends Variable> int asLit(Decision<V> decision) {
+        if (decision instanceof IntDecision) {
+            IntDecision id = (IntDecision) decision;
+            return asLit(
+                nogoodStealer.getById(id.getDecisionVariable(), png.getModel()), id.getDecOp(),
+                id.getDecisionValue()
+            );
+        } else if (decision instanceof SetDecision) {
+            SetDecision id = (SetDecision) decision;
+            return asLit(
+                nogoodStealer.getById(id.getDecisionVariable(), png.getModel()), id.getDecOp(),
+                id.getDecisionValue()
+            );
```

Comment:
This code could be easier by transfering `getDecOp()` to the `Decision<E>` class, where it would return a `DecisionOperator<E>`.

Reply:
I'm not sure this would be possible considering that `RealDecision` does not rely on `DecisionOperator` since it requires `value` to be a `double` and not an `int` as needed for `IntDecision` and `SetDecision`.
The same goes for `RootDecision`, but that's actually not a problem.

Reply:
This should be evaluated in another PR, because I'm confident with the performance when `DecisionOperator` is generified (and its consequence on `Decision`, `Search` and strategies)